### ISSUE TITLE
ConfigProviderResolver#setInstance is an implementation leak and should be dropped from the API

### ISSUE THREAD

**Issue by rmannibucau** [eclipse/microprofile-config#364]
See https://github.com/eclipse/microprofile-open-api/issues/224, this kind of method don't belong to the API and is only issues promises so should disappear from any coming release

**Comment by Emily-Jiang** [comment]
I disagree @rmannibucau ! It has usage and it is part of SPI, which is to provide a hook point for implementations.

**Comment by rmannibucau** [comment]
@Emily-Jiang well, as a fact no impl need this method so this is not needed. Don't forget the api jar must only contain user facing api and spi and not vendor ones.

**Comment by jmesnil** [comment]
@OndroMih is it still relevant?

**Comment by jeanouii** [comment]
Kinda agree with Romain on this one. This looks like an implementation detail but not something we want on a public API

**Comment by dmlloyd** [comment]
Perhaps https://osgi.org/specification/osgi.cmpn/7.0.0/service.loader.html is relevant here.

**Comment by smillidge** [comment]
-1 we are currently using this api

**Comment by dmlloyd** [comment]
I will point out that relying on this API is hazardous: it introduces a window of time where the wrong config provider resolver is available before the right one is set.  This can lead to leakage and even potential locking problems, depending on the config sources involved.  So users have to be very careful of this fact.  It should, at the very least, be well-documented.

We were using this API in Quarkus but the above problem hit us multiple times.  To avoid it, we've introduced an SPI in our implementation (SmallRye Config) which allows the process to be customized at a different point, so that there is no window of time where the wrong configuration might be visible/constructed.

**Comment by smillidge** [comment]
We are using it as an SPI during boot of Payara.

**Comment by rmannibucau** [comment]
Let make it simple. Standard for spec is to use a ServiceLoader.
For MP, only config needs a ServiceLoader or CDI events in extensions - all other specs can use CDI beans.
So let's fix this design bug (it is just not deterministic and breaks other methods to expose it since release can be broken then).

Guess no vendor is really blocked by the removal or full deprecation of that method since replacement is trivial.

**Comment by dmlloyd** [comment]
Just as a note, it is important that this specification continue to be usable outside of CDI.

**Comment by rmannibucau** [comment]
To be usable outside of cdi contextuality (Config.getConfig()) is "forbidden" cause you must respect your app contextuality and configuration (your own ioc or so, it is as anything or spring web for example). So this is completely fine.
